### PR TITLE
[WIP] Parse Typescript definition to get basic type information for decorators

### PR DESCRIPTION
**Pull request by timotheeguerin**
This is a work in progress, lots of things probably will need to be moved around. There also seems to be a massive performance problem as it is right now, editor is much slower
image

TODO: 
- [ ] Check performance issue
- [ ] Create ts compiler host using Cadl host
- [ ] Implement spread
- [ ] Include types for projection functions?
- [ ] Remove the create decoratorDefinition helper and only use internally. Maybe replace with a manual `setDecoratorSignature(fn: Funciton, signature: Decoratorsignature)`

### CODE REVIEW

File: packages/compiler/lib/decorators.ts [hunk 1]
```diff
@@ -87,7 +87,12 @@ const docsKey = Symbol("docs");
  *
  * @doc can be specified on any language element -- a model, an operation, a namespace, etc.
  */
-export function $doc(context: DecoratorContext, target: Type, text: string, sourceObject: Type) {
+export function $doc(
+  context: DecoratorContext,
+  target: NamespaceType | OperationType, // TODO-TIM revert back, for testing only
```
Comment (timotheeguerin) [inline packages/compiler/lib/decorators.ts:92]: todo remove when done testing

**Comment by bterlson** [comment]
It looks like we're probably creating the TS compiler host from scratch on each keypress. Decent performance can likely be gained fairly trivially (tm)  by caching the compiler instance re-using existing source files when they are unchanged.

**Comment by timotheeguerin** [comment]
The other thing seems to be the in the TS CompilerHost API everything seems sync which is incompatible with our CompilerHost using async

**Comment by markcowl** [comment]
@timotheeguerin What do we need to do to make progress on this PR?  Or should we close it?

**Comment by timotheeguerin** [comment]
> @timotheeguerin What do we need to do to make progress on this PR? Or should we close it?

Was going to talke about this PR during the design meeting and the issues with it.

**Comment by timotheeguerin** [comment]
Closing as this is not going to be the approach we go with moving forward

**Pull request closed without merging**